graylog2-server/src/main/java/org/graylog2/shared/rest/resources/ProxiedResource.java

```diff
@@ -112,4 +114,82 @@ protected ProxiedResource(@Context HttpHeaders httpHeaders,
             }
         };
     }
+
+    /**
+     * Execute the given remote interface function on the master node.
+     * <p>
+     * This is used to forward an API request to the master node. It is useful in situations where an API call can
+     * only be executed on the master node.
+     * <p>
+     * The returned {@link MasterResponse} object is constructed from the remote response's status code and body.
+     */
+    protected <RemoteInterfaceType, RemoteCallResponseType> MasterResponse<RemoteCallResponseType> requestOnMaster(
+            Function<RemoteInterfaceType, Call<RemoteCallResponseType>> remoteInterfaceFunction,
+            Function<String, Optional<RemoteInterfaceType>> remoteInterfaceProvider
+    ) throws IOException {
+        final Node masterNode = nodeService.allActive().values().stream()
+                .filter(Node::isMaster)
+                .findFirst()
+                .orElseThrow(() -> new IllegalStateException("No active master node found"));
+
+        final RemoteInterfaceType remoteInterfaceType = remoteInterfaceProvider.apply(masterNode.getNodeId())
+                .orElseThrow(() -> new IllegalStateException("Master node " + masterNode.getNodeId() + " not found"));
+
+        final Call<RemoteCallResponseType> call = remoteInterfaceFunction.apply(remoteInterfaceType);
+        final Response<RemoteCallResponseType> response = call.execute();
+
+        final byte[] errorBody = response.errorBody() == null ? null : response.errorBody().bytes();
+
+        return MasterResponse.create(response.isSuccessful(), response.code(), response.body(), errorBody);
```

Comment:
I do not really see the point of using the `MasterResponse` class here. Do we expect the requests to come up with a negative response regularly? If not, we should just throw exceptions wrapping the status. This way we also force the consumer to handle those instead of relying that the response object is parsed.

Reply:
We are using the `requestOnMaster()` method mostly in cases where we want to transparently proxy the response of the master request to the calling client. My first implementation returned a `javax.ws.rs.core.Response` here but I found that to be not flexible enough when using it because the caller might want to access the typed response object. But I also didn't want to return the `retrofit2.Response` here because the callers would have to write some code all the time to build a `Response` depending on the success status of the request.

This is how you can use the `MasterResponse` in a resource:

```java
final MasterResponse<T> response = requestOnMaster(...);

return Response.status(response.code()).entity(response.entity()).build();
```
The `MasterResponse#entity()` method returns the correct value depending on the success status of the response.

If we would return the `retrofit2.Response` here, you would have to check if we are dealing with an error or a success to select the correct value for the `Response#entity()`.

I didn't want to throw an exception here because I want to be able to return the error response from the master as it is without wrapping it in some exception. I guess we could also do that with a custom exception implementation but that would also require us to write some code.

So I decided to use a small `MasterResponse` object.